### PR TITLE
muon scale variation from smearing weights calculated on massfit stats unc

### PR DESCRIPTION
**Pull request by tyjyang**
This is a WIP PR, df.Define is not happy with my jpsi_unc_helper, and I need to figure out why.

### CODE REVIEW

**Comment by kdlong** [comment]
I'm starting to look at this now, but first I would say that the error messages can be very misleading, I think it's more likely a compilation error than an error in the define

File: scripts/histmakers/mw_with_mu_eta_pt.py [hunk 1]
```diff
@@ -306,27 +318,51 @@ def build_graph(df, dataset):
...
+                            "goodMuons_eta0_gen",
+                            "goodMuons_pt0_gen",
+                            "goodMuons_qop0_gen_smeared",
+                            "goodMuons_phi0_gen_smeared",
+                            "goodMuons_charge0_gen_smeared",
+                            "goodMuons_eta0_gen_smeared",
+                            "goodMuons_pt0_gen_smeared",
+                            "covMat_goodGenMuons0",
+                            "nominal_weight",
+                            True,
```
[the remark this answers is not in the document]
Comment (tyjyang) [reply scripts/histmakers/mw_with_mu_eta_pt.py:337]: Thanks Kenneth! Yes, I had the same suspicion, will change it to how it was done in the calibration uncertainties helper :)

**Comment by tyjyang** [comment]
The production of muon scale variation histograms from smearing weights calculated from stat. unc. on jpsi massfits are now implemented in the mw script. It is able to run through and produce the pickle file.

Validation is ongoing for the scale variation histogram produced. Specifically, I am checking if the variation histograms summed in quadrature in the pt bins can reproduce the shape of the massweights and manual shifts.

**Comment by tyjyang** [comment]
@kdlong @bendavid PR ready for the fully working RECO kinematics

File: wremnants/include/muon_validation.h [hunk 2]
```diff
@@ -56,23 +52,128 @@ class JpsiCorrectionsHelper {
...
+
+    // variadic templated bin lookup
+    template<typename... Xs>
+    const tensor_t &get_tensor(const Xs&... xs) {
+        return get_tensor_impl(std::index_sequence_for<Xs...>{}, xs...);
+    }
+
+    // for smearing weights derived from propagating uncs on A, e, M to uncs on qop
+    out_tensor_t operator() (
+        double genQop, float genPhi, int genCharge, float genEta, float genPt,
```
Comment (kdlong) [inline wremnants/include/muon_validation.h:100]: Could get charge from the sign of qop and slightly reduce the number of arguments you need

File: wremnants/include/muon_validation.h [hunk 3]
```diff
@@ -56,23 +52,128 @@ class JpsiCorrectionsHelper {
...
+        double recoQop, float recoPhi, int recoCharge, float recoEta, float recoPt,
+        const RVec<float> &cov, // for sigma on the Gaussian
+        double nominal_weight, bool abQop = true, bool fullParam = false
```
Comment (kdlong) [inline wremnants/include/muon_validation.h:103]: What does abQop stand for / do?

File: wremnants/include/muon_validation.h [hunk 4]
```diff
@@ -56,23 +52,128 @@ class JpsiCorrectionsHelper {
...
+    ) {
+        Eigen::Vector3d parms(
+            (abQop? recoQop : calculateQop(recoPt, recoEta, recoCharge)),
```
Comment (kdlong) [inline wremnants/include/muon_validation.h:106]: Why do you ever want to recalculate qop if you also pass it in? Seems like you would either always calculate it (to be sure it's consistent) or assume that the user passes in consistent values

File: wremnants/include/muon_validation.h (hunk 3 above)
Comment (tyjyang) [reply wremnants/include/muon_validation.h:103]: abQop tells whether the gen smearing is done on gen_pt or gen_qop

File: wremnants/include/muon_validation.h (hunk 4 above)
Comment (tyjyang) [reply wremnants/include/muon_validation.h:106]: this is also from whether you smear gen_pt or gen_qop

File: wremnants/include/muon_validation.h (hunk 2 above)
Comment (tyjyang) [reply wremnants/include/muon_validation.h]: I got rid of genCharge and genPt, which are only used to calculate genQop; and like you said, we don't really need it. The calculation is only needed for RECO which is smeared GEN at the moment.

File: wremnants/muon_calibration.py [hunk 5]
```diff
@@ -240,28 +241,30 @@ def define_uncrct_reco_muon_kinematics(df, kinematic_vars = ["pt", "eta", "phi",
...
-            proc_hists['nominal_gen_smeared'].project(*proj_axes))
-        msv_sw_gen_smear = [
-            (proc_hists['muonScaleSyst_responseWeights_gensmear'][...,0,0].project(*proj_axes)),
-            (proc_hists['muonScaleSyst_responseWeights_gensmear'][...,0,1].project(*proj_axes))
-        ]
-        sw_per_bin_gen_smear = [hh.divideHists(x, nominal_gen_smear) for x in msv_sw_gen_smear]
-        nominal_reco = proc_hists['nominal'].project(*proj_axes)
-        msv_sw_reco = [hh.multiplyHists(nominal_reco, x) for x in sw_per_bin_gen_smear]
-        proc_hists['muonScaleSyst_responseWeights'] = hh.combineUpDownVarHists(*msv_sw_reco)
+        nominal_gen_smear = resultdict[proc]['output']['nominal_gen_smeared']
```
Comment (kdlong) [inline wremnants/muon_calibration.py:248]: Probably a good idea to have some checks that the histograms exist here, I've had it happen in the past that this code is out of sync and it fails right at the end before writing the output, which is a bit annoying if you're doing other things anyway. If anything is missing you could throw from here and catch it in the main function with a warning message

Comment (tyjyang) [reply wremnants/muon_calibration.py]: ✅

File: wremnants/include/muon_validation.h (hunk 4 above)
Comment (tyjyang) [reply wremnants/include/muon_validation.h]: got rid of abQop in the helper, figured we could just pass in an RDF column of qop calculated from smeared gen_pt instead.

**Pull request merged**